File: packages/ravi-os-swift-sdk/Sources/RaviSDK/RaviTransport.swift
import Foundation

public protocol RaviTransport: Sendable {
  func call<T: Decodable & Sendable>(
    groupSegments: [String],
    command: String,
    body: [String: RaviJSON],
    as type: T.Type
  ) async throws -> T

  func callBinary(
    groupSegments: [String],
    command: String,
    body: [String: RaviJSON]
  ) async throws -> RaviBinaryResponse
}

public struct RaviBinaryResponse: Sendable {
  public let data: Data
  public let contentType: String?
  public let statusCode: Int
  public let headers: [String: String]

  public init(data: Data, contentType: String?, statusCode: Int, headers: [String: String]) {
    self.data = data
    self.contentType = contentType
    self.statusCode = statusCode
    self.headers = headers
  }
}

File: packages/ravi-os-dart-sdk/lib/src/ravi_transport.dart
import 'ravi_json.dart';

abstract class RaviTransport {
  Future<T> callJson<T>({
    required List<String> groupSegments,
    required String command,
    required Map<String, RaviJson> body,
    required T Function(Object? json) decode,
  });

  Future<RaviBinaryResponse> callBinary({
    required List<String> groupSegments,
    required String command,
    required Map<String, RaviJson> body,
  });
}

class RaviBinaryResponse {
  const RaviBinaryResponse({
    required this.bytes,
    this.contentType,
    required this.statusCode,
    required this.headers,
  });

  final List<int> bytes;
  final String? contentType;
  final int statusCode;
  final Map<String, String> headers;
}
